Sources/MongoSwift/BSON/BsonEncoderNew.swift

```diff
@@ -0,0 +1,647 @@
+        assertCanEncodeNewValue()
+        self.storage.push(container: try self.box(value))
+    }
+}
+
+/// A private class wrapping a Swift array so we can pass it by reference for 
+/// encoder storage purposes. We use this rather than NSMutableArray because
+/// it allows us to preserve Swift type information. 
+private class MutableArray: BsonValue {
+
+    var bsonType: BsonType { return .array }
+
+    var array = [BsonValue?]()
+
+    fileprivate func add(_ value: BsonValue?) {
+        array.append(value)
+    }
+
+    var count: Int { return array.count }
+
+    /// Converts self to a `Document` where keys "0", "1", etc.
+    /// correspond to array indices. 
+    func asDocument() -> Document {
+        var doc = Document()
+        for (i, v) in array.enumerated() {
+            doc[String(i)] = v
+        }
+        return doc
+    }
+
+    func encode(to data: UnsafeMutablePointer<bson_t>, forKey key: String) throws {
+        try self.array.encode(to: data, forKey: key)
+    }
+
+    static func from(iter: inout bson_iter_t) -> BsonValue {
+        return [BsonValue].from(iter: &iter)
+    }
+
+    func insert(_ value: BsonValue?, at index: Int) {
+        self.array.insert(value, at: index)
+    }
+}
+
+/// A private class wrapping a Swift dictionary so we can pass it by reference
+/// for encoder storage purposes. We use this rather than NSMutableDictionary 
+/// because it allows us to preserve Swift type information.
+private class MutableDictionary: BsonValue {
+
+    var bsonType: BsonType { return .document }
+
+    // rather than using a dictionary, do this so we preserve key orders
+    var keys = [String]()
+    var values = [BsonValue?]()
+
+    subscript(key: String) -> BsonValue? {
+        get {
+            guard let index = keys.index(of: key) else { return nil }
+            return values[index]
+        }
+        set(newValue) {
+            keys.append(key)
+            values.append(newValue)
+        }
+    }
+
+    /// Converts self to a `Document` with equivalent key-value pairs.
+    func asDocument() -> Document {
+        var doc = Document()
+        for i in 0 ..< keys.count {
+            doc[keys[i]] = values[i]
+        }
+        return doc
+    }
+
+    func encode(to data: UnsafeMutablePointer<bson_t>, forKey key: String) throws {
+        try self.asDocument().encode(to: data, forKey: key)
+    }
+
+    static func from(iter: inout bson_iter_t) -> BsonValue {
+        return Document.from(iter: &iter)
+    }
+}
+
+private extension EncodingError {
+    static func _numberError<T: CodableNumber>(at path: [CodingKey], value: T) -> EncodingError {
+        let description = "Value \(String(describing: value)) of type \(type(of: value)) cannot be " +
+                            "exactly represented by a BSON number type (Int, Int32, Int64 or Double)."
+        return .invalidValue(value, Context(codingPath: path, debugDescription: description))
+    }
+}
+
+extension Document: Encodable {
```

Comment:
I think this should be moved to like `Document+Codable.swift` or something

Reply:
moved, and also pulled in the `extension Document: Decodable` code. note this required changing some types and properties from `private`/`fileprivate` to `internal`.